blog/2028-08-13-python-typing/index.mdx

```diff
@@ -0,0 +1,161 @@
+
+```python
+from weaviate.collections import Collection
+from weaviate.outputs.query import Object
+
+def search(collection: Collection, vector: str) -> list[Object]:
+    return collection.query.near_vector(vector).objects
+```
+
+Here, `mypy` will flag the error at the typing of the `.near_vector` method with the following errors: **`error:** Returning Any from function declared to return **"list[Object[Any, Any]]"**  [no-any-return]` and **`error:** No overload variant of **"near_vector"** of **"_NearVectorQuery"** matches argument type **"str"**  [call-overload]`. Signifying first that the type inference cannot detect the type of `collection.query.near_vector(vector).objects` due to the misconfiguration that `vector: str` here and second that the actual type of `vector: str` is not expected by the v4 client.
+
+Improvements like these have been made throughout the v4 Client codebase. As a result, we think you will benefit from fewer bugs, and write fewer bugs yourself.
+
+## Nothing's Perfect: Type Hinting's Limitations
+
+While we've seen the many benefits of type hinting, it's important to acknowledge that no system is without its limitations.
+
+All of the above is great for types that are known specifically at the static stage of the code's lifecycle, meaning before any runtime invocation. But what happens when we're writing code that depends on data coming from some third-party external source?
+
+When working with a database like Weaviate, this is what we end up doing most of the time. So how can we provide the client, and therefore the type checker, with the types of data that we have stored in Weaviate?
+
+*Generics have entered the chat*
+
+## Generics in Python
+
+In short, *generic types* or just *generics* (which is just a bit of jargon describing placeholder values like symbols in algebra: `F=ma` or `E=mc^2`) are types that are not known by client when we package it together and release to you but can be specified dynamically by you when writing your applications. They provide a way for the you, the user, to annotate your client with the data types that you expect to load into and query from Weaviate. So, if you have a collection that has a property like `Property(name="title", data_type=DataType.TEXT)` then you can annotate your collection so that accessing the `title` property of an object from that collection will be fully typed as so: `x: str = obj.properties["title"]`.
+
+The v4 Weaviate client API comes with full support for generics when defining collections and the types of the objects within the collections. But details on how we implemented it and some nifty tricks on getting the best out of it will come in a later post.
```

Comment:
WDYT about adding an emoji at the end of this so it's less sad that it won't come in this article?

Reply:
Rephrased to:

> We'd love to tell you more about how we implemented it and some nifty tricks on getting the best out of it. So please stay tuned for those details in a later post 😉.